Export Julia's LLVM passes using LLVM C API compatible functions.

maleadt opened the pull request:
Used for [CUDAnative.jl](https://github.com/JuliaGPU/CUDAnative.jl/blob/14f215aac8f524094d69b5185b2976a856dcbe86/src/jit.jl#L214-L230).

maleadt committed:
Export Julia's LLVM passes using LLVM C API compatible functions.

Used for external language implementations.

vtjnash commented:
Maybe would be easier to just eliminate these tbaa node arguments?

yuyichao commented:
> Maybe would be easier to just eliminate these tbaa node arguments?

Sounds good to me, I initially did this to make it as independent as possible but since I couldn't figure out a way to make the pass truly independent of the context it probably not worth doing this anymore...

maleadt committed:
Remove the MDNode arguments to Julia's LLVM passes.

We reconstruct the TBAA MDNodes in the pass' runOnModule functions.
This should be fine, as the node's first field uniquely identifies it.

maleadt commented:
Eliminated the tbaa node arguments, recreating them in the pass. Did so in `runOnModule`, at the cost of some extra verbosity, to have it working with multiple contexts.

Should be identical, but I guess it can't hurt:

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19024#issuecomment-254934925) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/05691e7399cda0d625839e6899c27b151c431361/c0feabd_vs_0faf8ce/report.md). cc @jrevels

maleadt commented:
OK, looks good. With Travis OSX being dead, I built and tested on the MIT macmini which went fine.

yuyichao commented on src/llvm-gcroot.cpp:
```diff
@@ -1175,6 +1176,10 @@ static void ensure_enter_function(Module &M)
 
 bool LowerGCFrame::runOnModule(Module &M)
 {
+    MDBuilder *mbuilder = new MDBuilder(M.getContext());
```
Should this be stack allocated instead?

maleadt replied:
Probably yeah, I copied without paying attention...

yuyichao reviewed (approved):
LGTM assuming you've check that the child node we create each time are the same one. (so that we are not creating a increasing numbers of them).

maleadt commented:
@yuyichao Yeah I verified, no new nodes are created. But even if they were, they should 'alias' I guess (`may be identified (uniqued) with other roots with the same name`).

I'll merge if the board turns green.

vtjnash commented on src/codegen.cpp:
```diff
@@ -5078,7 +5085,7 @@ static std::unique_ptr<Module> emit_function(jl_method_instance_t *lam, jl_code_
-static MDNode *tbaa_make_child( const char *name, MDNode *parent, bool isConstant=false )
+MDNode *tbaa_make_child(const char *name, MDNode *parent, bool isConstant=false)
 {
     MDNode *n = mbuilder->createTBAANode(name,parent,isConstant);
```
`MDBuilder mbuilder(parent->getContext())`

the `MDBuilder` class is just a single pointer (to a LLVMContext), it doesn't need to be a global

vtjnash commented:
LGTM, removing globals from codegen is a future cleanup action and doesn't need to be done now
